### ISSUE TITLE
[CI][docs] Auto-generate release notes.

### ISSUE THREAD

**Issue by JDarDagran** [OpenLineage/OpenLineage#3021]
### Motivation

As part of the effort to improve docs and their versioning having release notes automatically generated would be a nice feature.

### Description

https://pypi.org/project/towncrier/ might be a nice tool to restructure and automate release notes.

### Do you plan to make this contribution yourself?

- [X] I am interested in doing this work

**Comment by mobuchowski** [comment]
Would that work for other things than Python client at this point?

**Comment by dolfinus** [comment]
This works as follows:
* Contributor creates PR with some feature/bugfix/etc. This PR should include a file with human-readable description, like `docs/changes/next_release/1234.feature.rst`, where 1234 is PR id.
* Before publishing a release, repo maintainer runs `towncrier build` which is a CLI tool allowing to merge all these files into one release note file. It supports .rst format, and probably .md
* In the end, there is a file describing all release notes for the current release, which should be merged to the main branch.

So using `towncrier` does not really interfere with a language used in the repo, as it does not parse docstrings, commit messages - it just merges files.

**Comment by JDarDagran** [comment]
I read some more: `towncrier` would solve some issues while generating other. One is that it would force conributors to judge what is the type of change (maintainers could help ofc). The other is that we would need to force users to include the changelog with the PR (but does every PR need changelog?).

**Comment by dolfinus** [comment]
> One is that it would force conributors to judge what is the type of change (maintainers could help ofc)

Yep, towncrier can check that change type (`feature` from `1234.feature.rst` file name) is present in predefined list of change types in a config file.

> The other is that we would need to force users to include the changelog with the PR (but does every PR need changelog?).

In my project there is a Github Actions which checks that file mentioning current PR id exists in `next_release` folder. But there is a special label `skip-changelog`, which usually applied to changed not related to project source code, like CI updates.

**Comment by JDarDagran** [comment]
> In my project there is a Github Actions which checks that file mentioning current PR id exists in `next_release` folder. But there is a special label `skip-changelog`, which usually applied to changed not related to project source code, like CI updates.

Is the label automatically assigned or maintainers decide on this?

**Comment by dolfinus** [comment]
> Is the label automatically assigned or maintainers decide on this?

It's up to maintainers

**Comment by JDarDagran** [comment]
How is the GHA triggered then? Something like:
```yaml
on: 
  pull_request:
    types: [labeled, unlabeled]
 ```
 ?

**Comment by dolfinus** [comment]
Here is an example: https://github.com/MobileTeleSystems/onetl/blob/cb6187402201a8bf71740cbfad33711e16bc7084/.github/workflows/changelog.yml

**Comment by d-m-h** [comment]
There is also the concept of [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/). This is something that I have used before, and it helps to categorize and organise the change log.

**Comment by merobi-hub** [comment]
Do any of these tools support annotation of changes? We currently use a script that partially automates changelog updates by parsing PRs to add these (if contributors have provided them where the parser looks for them). If not, wouldn't the changelog be much terser? Would that be much of a loss?

**Comment by JDarDagran** [comment]
Closing this for now. Currently release is published manually by release manager in GitHub. This triggers GHA workflow to update release notes on the docs site as well.
